### PR TITLE
[WIP] Make chown tolerant of errors when doing single user mapping

### PR DESCRIPTION
**Pull request by kpelzel**
**Linked to this PR:** https://github.com/containers/storage/pull/358

### Problem:
When building a container as an unprivileged user and without uidmap tools, buildah will fail when chowning system files. This patch ignores those chown errors if buildah has fallen back to single user mapping mode.

### What this patch changes:
**Buildah side:** 
unshare.go checks if newuidmap and newgidmap is installed by running "which newuidmap". If uidmap tools are not installed it sets an environment variable to pass down to the child.

**containers/storage side:**
diff_unix.go passes in the hosts environment variables (including the single user mapping environment variable) to the "storage-applyLayer" command. This allows checkChownErr in idtools.go to check if it needs to ignore the chown errors.

### Notes:
I'm not completely sure if using environment variables is the best way to get this information down to checkChownErr, but it seemed to work best in terms minimal code change. I would love to hear feedback on better solutions.

### CODE REVIEW

**Comment by giuseppe** [comment]
bot, add author to whitelist

File: pkg/unshare/unshare.go [hunk 1]
```diff
@@ -558,3 +586,25 @@ func ParseIDMappings(uidmap, gidmap []string) ([]idtools.IDMap, []idtools.IDMap,
 	}
 	return uid, gid, nil
 }
+
+func checkGIDMap() error {
+	cmd := exec.Command("which", "newgidmap")
```
Comment (giuseppe) [inline pkg/unshare/unshare.go]: I think we want to use `exec.LookPath` here instead of running another process

File: pkg/unshare/unshare.go [hunk 2]
```diff
@@ -558,3 +586,25 @@ func ParseIDMappings(uidmap, gidmap []string) ([]idtools.IDMap, []idtools.IDMap,
...
+	cmd.Stderr = nil
+	err := cmd.Run()
+	if err != nil {
+		return errors.Wrapf(err, "command 'which newgidmap' failed. Is newgidmap in $PATH?")
+	}
+	return nil
+}
+
+func checkUIDMap() error {
+	cmd := exec.Command("which", "newuidmap")
```
Comment (giuseppe) [inline pkg/unshare/unshare.go]: same here

**Comment by kpelzel** [comment]
In the most recent commit I added a "single-user-mapping" global flag to buildah. This will set the environment variable that tells idtools to ignore chown errors. I didn't know what to name the flag so feedback is welcome.

I also moved most of the uidmap checking logic into MaybeReexecUsingUserNamespace because I thought it made the code more readable, but let me know if that's a bad idea.

**Comment by rhatdan** [comment]
Changing to WIP flag, to make sure this does not get merged until we get the containers/storage patches merged and figure out how this should interact with containers/storage.

**Comment by kpelzel** [comment]
Closing as this issue is resolved in https://github.com/containers/buildah/pull/1749

**Comment by rhatdan** [comment]
@kpelzel Once we get this support into podman 1.5, we would love to have further feedback on what you would need to make podman a good solution for HPC environments.

**Pull request closed without merging**